---
title: "dada2 to omeClust and MaAsLin2 pipeline"
author: "Clark Gaylord (cgaylord@gwu.edu)"
date: "12/16/2021"
output: html_document
---

We wish to give a demonstration of using the dada2 denoising framework as an input to both omeClust and MaAsLin2 using data from Song, et al. Note that omeClust is an external application; though we will run it from this script, the output will go to a directory. MaAsLin2 also writes its output to a directory.

Our code is a simplified version of our analysis for the purposes of this tutorial.

The libraries we use are:
lubridate
tools
dada2
ggplot2
bioDist
Maaslin2

These are assumed to have been installed and available as libraries.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# now with timing!
TIMING = F
DEBUG  = F
SINK   = F
sink_file = "pipeline.out"

# Where are the .fastq.gz files?
data_directory <- "./ENAfiles"
# Choose aa arbitrary file to view the quality scores for
example_sequence_file <- "ERR3947849.fastq.gz"
metadata_file <- "./fecal_onemonth-metadata.txt"
omeClust_directory <- "ome_results"
maas_directory <- "maaslin_infant-sqrt"
print(paste("Output will go to", omeClust_directory, "and", maas_directory))

if (SINK) {
  if (file.exists(sink_file)) {
    file.rename(sink_file, paste0(sink_file,".bak"))
  }
  sink(file = sink_file, split = T)
}

# Uncomment any installs that are required
# These should be conditionals instead
# dada2 probably needs to be installed from source
# install.packages("dada2")

# library(devtools)
# devtools::install_github("benjjneb/dada2", ref="v1.18")
# change the ref argument to get other versions

# BiocManager::install("bioDist")
# BiocManager::install("Maaslin2")

# Ended up not using phyloseq
# BiocManager::install("phyloseq")

library(lubridate)
library(tools)
# Converted to using a text metadata file
# library(readxl)
library(dada2)
library(ggplot2)
library(bioDist)
library(Maaslin2)
# We don't use phyloseq but it is a good choice to explore with dada2
# library(phyloseq)

if (TIMING) now()
```


The first step in the dada2 pipeline is to trim and filter the low quality sequences from the data. We choose one of the samples arbitrarily to visualize its quality. Based on this (or any other) plots, you may choose to trim differently or to use a minimum quality score as a threshold, but we will trim the leftmost 25 and rightmost 10 bases from our samples.

```{r qualityPlot, echo=FALSE}
plotQualityProfile(paste0(data_directory, "/", example_sequence_file))

# raw data files are compressed
Infant_ALL_seq_files <- list.files(
    data_directory,
    pattern = "fastq.gz",
    full.names = T)

# Eyeballing the quality plots we choose to trim the ends
# One can also add filtering on quality score (say minQ=15 or 20)
# Filtered fastq files are placed into FilteredENAfiles directory
if (TIMING) now()

filtered_directory <- paste0(data_directory, ".filtered")
#
# This takes several minutes. On my Mac with 138 sequences it is nearly 10 minutes.
# Not bad, but go makea  cup of coffee.
#
filtFastqReads <- filterAndTrim(
    data_directory,
    filtered_directory,
    trimLeft = 25,
    trimRight = 10)

# Remove the .fastq.gz
# This is fragile, relying on the "double extension"
# by trying to avoid an ugly string search/replace got something worse
accessionNames <- file_path_sans_ext(file_path_sans_ext(basename(Infant_ALL_seq_files)))
row.names(filtFastqReads) <- accessionNames
# get rid of this ugliness
# substr(row.names(filtFastqReads), start=1, stop=10)
```

Some summarization of the filtered fastq reads

```{r summaryReads, echo=TRUE}
head(filtFastqReads)
summary(filtFastqReads)
print(paste(
    "Total reads ",
    sum(filtFastqReads[,1]),
    "and after filter",
    sum(filtFastqReads[,2])))
print("The proportion of reads retained is")
sum( filtFastqReads[,2]) / sum(filtFastqReads[,1] )
```

```{r plotSummaryReads, echo=FALSE}
plot(filtFastqReads[,2]~filtFastqReads[,1],
     main = "Filtered Reads",
     ylab = "Total sequence filtered",
     xlab = "Total sequence raw")
# Add the reference line y=x
abline(a=0, b=1)

# The axes should be more dynamically formatted
plot(density(filtFastqReads[,2], adj=0.5),
     main = "Density estimate of number of reads per sample",
     xaxt = "n", yaxt = "n")
axis(side = 1,
     labels = c("0", "100,000", "200,000", "300,000", "400,000"),
     at = c(0,100000,200000,300000,400000))
rug(filtFastqReads[,2])

if (TIMING) now()

Infant_ALL_filt_files <- list.files(
    filtered_directory,
    pattern = "ERR",
    full.names = T)
```

The dada algorthim uses an estimated error function as part of its input
```{r errorRates, echo=FALSE}
errRates <- learnErrors(Infant_ALL_filt_files, multithread = T)
plotErrors(errRates, nominalQ = T)

#
# The final sequence identification
#
if (TIMING) now()
```

And finally, dada2 denoises the sequences so that we can have a matrix of features (the amplicon sequences themselves) from each sample.
```{r dadaRun, echo=TRUE}
dada_Infant <- dada(
    Infant_ALL_filt_files,
    err = errRates,
    multithread = TRUE)
```

```{r sequenceSetup, echo=FALSE}
if (TIMING) now()

seqTable_Infant <- makeSequenceTable(dada_Infant)
row.names(seqTable_Infant) <- accessionNames
```

How many sequences came out of the denoising?
```{r denoisingOutput, echo=TRUE}
dim(seqTable_Infant)
table(nchar(getSequences(seqTable_Infant)))
```

Now to remove chimeras, we are left with our features to analyze:
```{r filterChimeras, echo=TRUE}
seqTable_Infant.nochim <- removeBimeraDenovo(
    seqTable_Infant,
    method = "consensus",
    multithread = TRUE,
    verbose = TRUE)

dim(seqTable_Infant.nochim)
table(nchar(getSequences(seqTable_Infant.nochim)))
```

Let's look at the abundance of a chosen sequence
```{r plotAbundance, echo=FALSE}

seqTable_Infant.nochim.df <- as.data.frame(seqTable_Infant.nochim)

# Now what would be really cool is to randomly select a couple
# sequences (or even just one), dynamically construct the plot
# and even do the BLAST query to get the best genus match.
# But nah.
#
# Create an example abundance plot
exampleSequence <- colnames(seqTable_Infant.nochim.df[11])
print(paste("Example sequence is ", exampleSequence))
subTitle <- paste("Nonzero abundance (",sum(seqTable_Infant.nochim.df[11]>0),")")
boxplot(sqrt(seqTable_Infant.nochim.df[11])[seqTable_Infant.nochim.df[11]>0],
        main="Example Bifidobacterium Abundance", sub = subTitle, ylab="sqrt(count)")
```

How many reads were filtered from each step of the pipeline (total number of reads and proportion from the initial set)?
```{r filterSummary, echo=FALSE}

getN <- function(x) sum(getUniques(x))
track <- cbind(
    filtFastqReads,
    sapply(dada_Infant, getN), 
	# we have single-ended reads; do these for double-ended
    # sapply(dadaRs, getN),
    # sapply(mergers, getN),
    rowSums(seqTable_Infant.nochim))
colnames(track) <- c(
    "input",
    "filtered",
    "denoisedF",
    # "denoisedR",
    # "merged",
    "nonchim")
rownames(track) <- accessionNames
head(track)
total_track <- colSums(track)
prop_track <- total_track/total_track[1]
total_track
prop_track

if (TIMING) now()
```

```{r taxaComments, echo=FALSE}

# We don't do taxonomy -- commenting out for reader's benefit
# taxa <- assignTaxonomy(
#     seqTable_Infant.nochim,
#     "silva_nr_v132_train_set.fa.gz",
#     multithread=TRUE)
# with 16S V4, there is probably no point to addSpecies but
# it is included in the dada2 tutorial
# if (TIMING) now()

# Note at this time we don't actually do anything with the taxonomy
# You could save yourself a little time by removing this.
# taxa <- addSpecies(
#     taxa,
#     "silva_species_assignment_v132.fa.gz")
# taxa.df <- as.data.frame(taxa)
# should we chuck sequences with <50 bases? Unusual sequence lengths?

# sequences <- (data.frame(cbind(
#     rownames(taxa.df),
#     as.numeric(nchar(rownames(taxa.df))))))
# colnames(sequences) <- c('sequence','length')
# sequences.df <- as.data.frame(sequences)

#
# Let's fold the metadata into the mix.
# There is much wrangling that has to happen here -- do what you need
# munge your metadata
# Ultimately omeClust and MaAsLin2 will need relatively simple arrays
#
```

```{r metadataAndDistance, echo=FALSE}
metadata <- read.delim("fecal_onemonth-metadata.txt")
# Metadata_Baby_Seeding_all_samples_final <-
#     read_excel("Metadata_Baby_Seeding_all_samples_final.xlsx",
#                sheet = "Possible")
metadata.sorted <-
    metadata[order(metadata$ena_accession),]
# The cool kids would use tidyverse
metadata.reorder <- as.matrix(cbind(
    metadata.sorted$ena_accession,
    metadata.sorted$baby_sex,
    metadata.sorted$birth_mode))
colnames(metadata.reorder) <- c(
    "ena_accession",
    "baby_sex",
    "birth_mode")
birth_mode <- as.data.frame(
    metadata.reorder[,3])
colnames(birth_mode) <- "birth_mode"
row.names(birth_mode) <- metadata.reorder[,1]

dist_seqTable_Infant.nochim <- as.matrix(dist(seqTable_Infant.nochim))

# Now we write out the data that omeClust will need
write.table(
    metadata.reorder,
    file="ome_infant_metadata.txt",
    sep="\t",
    row.names = F,
    quote = F)
write.table(
    dist_seqTable_Infant.nochim,
    file="ome_infant_distance.txt",
    row.names = T,
    col.names = T,
    sep="\t",
    quote=F)


# How to run omeClust (it isn't an R package)
#  omeClust
#      -i ome_infant_distance.txt \
#      -o ome_results \
#      --metadata ome_infant_metadata.txt \
#      --plot
#
# Wouldn't it be cool to go get the omeClust results and insert them here?
# That would sure make knitting a lot more complete.
if (TIMING) now()
```

Calling omeClust
```{r omeClustCall, echo=TRUE}
system(paste("omeClust -i ome_infant_distance.txt -o", omeClust_directory,"--metadata ome_infant_metadata.txt --plot"))
```

```{r junk1, echo=FALSE}

if (TIMING) now()

```

And now calling Maaslin2 (output is huge, so change this to echo=TRUE if you really want it)
```{r maasCall, echo=FALSE}
Maaslin2(
    sqrt(seqTable_Infant.nochim.df),
    birth_mode,
    output=maas_directory,
    reference = c("birth_mode", "CS"))
```

Now you can review the output in the respective directories.

```{r junk2, echo=FALSE}
if (TIMING) now()
if (SINK) sink()
```

Thank you for playing.